### PR TITLE
Fix PostgreSQL `Cidr#change?` to compare with address prefix

### LINKED ISSUE #51582: [ActiveRecord] PostgreSQL Adapter skips update on CIDR column when only netmask is changed
### Steps to reproduce

I cannot provide an executable test case of the exact problem since PostgreSQL cannot run in-memory easily, but I can demonstrate the underlying problem.
When I have a CIDR column in a PostgreSQL database with value `1.1.1.0/24` and I change it to `1.1.1.0/25`, it will not be updated on save.

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails"
end

require "active_record"
require "active_record/connection_adapters/postgresql/oid"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.logger = Logger.new(STDOUT)

class BugTest < Minitest::Test
  def test_netmask_change
    type = ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Cidr.new()
    old_value = type.cast_value("1.1.1.0/24")

    new_str = "1.1.1.0/25"
    new_value = type.cast_value(new_str)

    assert type.changed?(old_value, new_value, new_str)
  end
end
```

### Expected behavior
When `old_value` is in any way different from `new_value`, then `changed?` should return `true`.

### Actual behavior
`old_value` (1.1.1.0/24) is treated as the same as `new_value` (1.1.1.0/25) which is obviously not the case. 

### Further notes
The cause of this problem is, that `IPAddr#==` ignores the netmask for its comparison.
To solve the problem, it is necessary to overwrite `ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Cidr#changed?` so it does not use `IPAddr#==` anymore. `IPAddr#eql?` is suitable for this, but may be overkill. There is an open issue addressing that `IPAddr#==` does not work as expected https://github.com/ruby/ipaddr/issues/21

It is possible to monkey-patch `ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Cidr` to work as expected:
```
class ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Cidr
  def changed?(old_value, new_value, _new_value_before_type_cast)
    return old_value.nil? != new_value.nil? || (!old_value.nil? && !old_value.eql?(new_value))
  end
end
```

### System configuration
**Rails version**: 7.1.3.2

**Ruby version**: 3.3.0

### PR DESCRIPTION
**Pull request by taketo1113**
### Motivation / Background

This Pull Request has been created because fixes #51582

### Detail

This Pull Request changes to override`ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Cidr#changed?`.
It added to compare address prefix of IPAddr class.

### Additional information

I’m proposing to fix `IPAddr#==` comparing address prefix for the cause of this problem.
https://github.com/ruby/ipaddr/pull/69

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb [hunk 1]
```diff
@@ -28,6 +28,11 @@ def serialize(value)
             end
           end
 
+          def changed?(old_value, new_value, _new_value_before_type_cast)
+            old_value.class != new_value.class ||
```
Comment (yahonda) [inline activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb:32]: I'm wondering if it needs `old_value.class != new_value.class` here. I assume it comes from https://github.com/rails/rails/commit/dbdd68b64b781ce3c3363ffbc223730138c49a33 , Cidr `cast_value` method looks like taking care of it https://github.com/rails/rails/blob/734719d8af033062cb7088a1353be6276659e3c1/activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb#L31-L43

Comment (taketo1113) [reply activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb]: @yahonda Sure I think too this is not necessary.
I have removed `old_value.class != new_value.class`.

File: activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb [hunk 2]
```diff
@@ -28,6 +28,10 @@ def serialize(value)
             end
           end
 
+          def changed?(old_value, new_value, _new_value_before_type_cast)
+            new_value != old_value || new_value.prefix != old_value.prefix
```
Comment (yahonda) [inline activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb:32]:
```suggestion
            super || new_value.prefix != old_value.prefix
```

Comment (yahonda) [reply activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb:32]: I think using `super` makes Cidr#changed? requirement clear. https://github.com/rails/rails/blob/734719d8af033062cb7088a1353be6276659e3c1/activemodel/lib/active_model/type/value.rb#L84-L86

Comment (yahonda) [reply activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb:32]: I just prefer this to keep the argument order inside the method. old_value, then new_value.
```ruby
super || old_value.prefix != new_value.prefix
```

Comment (taketo1113) [reply activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb]: Thank you for your advices.
It has fixed to adopt `super` and keeping the argument order.

File: activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb [hunk 3]
```diff
@@ -28,6 +28,10 @@ def serialize(value)
             end
           end
```
Comment (justinko) [inline activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb:30]:
```suggestion
TODO: Remove when IPAddr#== compares IPAddr#prefix
```

Comment (taketo1113) [reply activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb:30]: Added todo comment to remove method and issues url.

**Review by yahonda — APPROVED** [review]

**Comment by yahonda** [comment]
Backported to the `7-1-stable` branch via https://github.com/rails/rails/commit/887210f8321d78ba9cdba9368a338c9411632f8b

**Comment by vipulnsward** [comment]
@yahonda This needs a changelog as its a change in behavior?

**Comment by yahonda** [comment]
Let me add a change log entry for this change. Thanks for the heads-up.

**Comment by yahonda** [comment]
Opened https://github.com/rails/rails/pull/51742 to add a changelog entry for this change.

**Comment by chaadow** [comment]
@yahonda @taketo1113 I'm on Rails 7-1 stable. and this makes my production app crash under certain circumstances. 
image

when `#save` is called on an activerecord model. here is the stack trace: 
image

**Comment by chaadow** [comment]
( I do not have a CIDR column in any of my tables)

**Comment by chaadow** [comment]
Honestly I wish I could reproduce the bug (it's in production) but it's so mystical. all I can say is, it's related to this PR and it was done on a "ActiveRecord#save" call. on an **unpersisted instance**

(probably the reason why `old_value.prefix` is nil ?)

but what's mystical to me is why it goes through the `changed?` call here even though I have 0 CIDR columns

**Comment by chaadow** [comment]
I feel the suggestion in #51582 here would fix my case:
```ruby
class ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Cidr
  def changed?(old_value, new_value, _new_value_before_type_cast)
     old_value.nil? != new_value.nil? || (!old_value.nil? && !old_value.eql?(new_value))
  end
end
```

or at least the part that checks that it's not nil

**Comment by chaadow** [comment]
@taketo1113 @yahonda I finally managed to recreate the issue

It's due to the fact **that `inet` inherits from CIDR**. 

this part helped me get less confused and better investigate and reproduce the issue

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  # If you want to test against edge Rails replace the previous line with this:
  gem "rails", github: "rails/rails", branch: "main"

  gem "pg"
end

require "active_record"
require "minitest/autorun"
require "logger"

db_config = {
  adapter: 'postgresql',
  database: 'inet_cidr_test',
  host: 'localhost', # Change this to your PostgreSQL host
  port: 5432,
  username: 'postgres', # Change this to your PostgreSQL username
  password: 'postgres'  # Change this to your PostgreSQL password
}

ActiveRecord::Base.establish_connection(db_config.except(:database))
ActiveRecord::Base.connection.drop_database(db_config[:database]) rescue nil
ActiveRecord::Base.connection.create_database(db_config[:database])

ActiveRecord::Base.establish_connection(db_config)
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :users, force: true do |t|
    t.inet :ip_address
  end
end

class User < ActiveRecord::Base
end


class BugTest < Minitest::Test
  def test_ip_address_with_empty_string
    User.create!(ip_address: '')
  end
end
```

**Comment by carlosantoniodasilva** [comment]
@chaadow if you haven't already, please open up a separate/new issue to help track this.

**Comment by taketo1113** [comment]
I confirmed the reproduce issue and created PR (#51758)

**Pull request merged**